Sources/Controllers/2.0/OARearrangeCustomFiltersViewController.mm

```diff
@@ -0,0 +1,453 @@
+//
+//  OARearrangeCustomFiltersViewController.mm
+//  OsmAnd
+//
+// Created by Skalii Dmitrii on 19.04.2021.
+// Copyright (c) 2021 OsmAnd. All rights reserved.
+//
+
+#import "OARearrangeCustomFiltersViewController.h"
+#import "OAPOIFiltersHelper.h"
+#import "Localization.h"
+#import "OADeleteButtonTableViewCell.h"
+#import "OAColors.h"
+#import "OAQuickSearchHelper.h"
+#import "OAButtonRightIconCell.h"
+#import "OAAppSettings.h"
+
+#define kAllFiltersSection 0
+#define kHiddenFiltersSection 1
+#define kActionsSection 2
+#define kHideButtonCell @"OADeleteButtonTableViewCell"
+#define kButtonRightIconCell @"OAButtonRightIconCell"
+#define kHeaderViewFont [UIFont systemFontOfSize:15.0]
+
+typedef void(^OAActionButtonOnClick)(id sender);
+
+@interface OAEditFilterItem : NSObject
+
+@property (nonatomic) int order;
+@property (nonatomic) OAPOIUIFilter *filter;
+
+- (instancetype) initWithFilter:(OAPOIUIFilter *)filter;
+
+@end
+
+@implementation OAEditFilterItem
+
+- (instancetype) initWithFilter:(OAPOIUIFilter *)filter
+{
+    self = [super init];
+    if (self) {
+        _filter = filter;
+        _order = filter.order;
+    }
+    return self;
+}
+
+@end
+
+@interface OAActionItem : NSObject
+
+@property (nonatomic) NSString *title;
+@property (nonatomic) UIImage *icon;
+@property (nonatomic) OAActionButtonOnClick onClickFunction;
+
+- (instancetype)initWithIcon:(UIImage *)icon title:(NSString *)title onClickFunction:(OAActionButtonOnClick)onClickFunction;
+- (void)onClick;
+
+@end
+
+@implementation OAActionItem
+
+- (instancetype)initWithIcon:(UIImage *)icon title:(NSString *)title onClickFunction:(OAActionButtonOnClick)onClickFunction
+{
+    self = [super init];
+    if (self) {
+        _title = title;
+        _icon = icon;
+        _onClickFunction = onClickFunction;
+    }
+    return self;
+}
+
+- (void)onClick
+{
+    self.onClickFunction(self);
+}
+
+@end
+
+@interface OARearrangeCustomFiltersViewController() <UITableViewDelegate, UITableViewDataSource>
+
+@property (weak, nonatomic) IBOutlet UIView *navBar;
+@property (weak, nonatomic) IBOutlet UITableView *tableView;
+@property (weak, nonatomic) IBOutlet UILabel *titleLabel;
+@property (weak, nonatomic) IBOutlet UIButton *cancelButton;
+@property (weak, nonatomic) IBOutlet UIButton *doneButton;
+
+@end
+
+@implementation OARearrangeCustomFiltersViewController
+{
+    OAAppSettings *_settings;
+    OAPOIFiltersHelper *_filtersHelper;
+    BOOL _isChanged;
+    BOOL _orderModified;
+    BOOL _hiddenModified;
+    BOOL _wasReset;
+
+    NSMutableArray<OAActionItem *> *_actionsItems;
```

Comment:
Generally, it is preferred to use immutable collections for items that are generated once and are never mutated. Please consider using immutable collections when possible.